lib/deep_delete.rb

```diff
@@ -0,0 +1,186 @@
+  IGNORE_CALLBACKS = {
+    # Storages#all_relationships(ems_metadata) - we're deleting all children anyway
+    "Storages"            => 1,
+    # EmsFolders#all_relationships(ems_metadata)
+    "EmsFolders"          => 1,
+    # EmsClusters#all_relationships(ems_metadata)
+    "EmsClusters"         => 1,
+    # ResourcePools#all_relationships(ems_metadata) - TODO: do we want ancestry here?
+    "ResourcePools"       => 1,
+    # MonitorManager => ExtManagementSystem
+    "ExtManagementSystem" => 1,
+    # ContainerManager#monitoring_manager (apply_orphan_strategy?)
+    # MonitorManager#endpoints: endpoint_destroyed (using this/destroy_queue instead of dependent => destroy)
+    # ContainerManager#persistent_volumes (?)
+    # InfraManager#orchestration_templates (check_not_in_use (before destroy)
+    # InfraManager#orchestration_stacks apply_orphan_strategy
+    # BlacklistedEvent#reload_all_server_settings, BlacklistedEvent#audit_deletion
+    "BlacklistedEvent"    => 2,
+    # EventStream.after_commit :emit_notifications, :on => :create
+    "EventStream"         => 1
+    # VmOrTemplate#apply_orphan_strategy (/via ancesty - needed to destroy tree)
+    # Relationship#apply_orphan_strategy (/via ancestry - needed to destroy tree)
+    # Endpoint#endpoint_destroyed (using this/destroy_queue instead of dependent => destroy)
+  }.freeze
+
+  IGNORE_CONSTRAINTS = {
+    # Hardware.disks is an order clause
+    "Hardware.disks"                          => true,
+    # MiqAlertStatus.miq_alert_status_actions is an order clause
+    "MiqAlertStatus.miq_alert_status_actions" => true,
+    # BinaryBlob.binary_blob_parts is an order clause
+    "BinaryBlob.binary_blob_parts"            => true
+  }.freeze
+
+  # true to recurse into associations that have no values and would be skipped
+  # Useful for developers to ensure all relationships could be reached
+  attr_accessor :visit_skips
+
+  def initialize(visit_skips: false)
+    @visit_skips = visit_skips
+  end
+
+  # @param scope the record(s) to be destroyed - only a single child class/type can be passed in
+  def self.delete(scope, **options)
+    if scope.kind_of?(ActiveRecord::Base)
+      # convert an individual record into a scope
+      scope = scope.class.where(:id => scope.id)
+    else
+      # update scope.klass to a more specific class.
+      # Ems child classes have additional relationships defined
+      record = scope.first
+      if record.nil?
+        _log.error("DeepDelete found no records")
+        return
+      end
+      scope = record.class.merge(scope)
+    end
+
+    name = scope.klass.name.split(":").last
+    _, timing = Benchmark.realtime_block(:deep_delete) do
+      new(**options).recursive_delete(scope, :name => name)
+    end
+    _log.info("Finished DeepDelete in #{timing[:deep_delete]} seconds")
+
+    self
+  end
+
+  def self.vacuum
+    ActiveRecord::Base.connection.execute("vacuum analyze")
+    self
+  end
+
+  ### this is the main method
```

Comment:
Can this doc string be improved?  It doesn't really tell what it does other than being the expected entrypoint.

Reply:
this is the only public method
It is probably not needed